File: erlang/pe012.erl
-module(pe012).
-export([start/0]).

max() -> 500.

start() ->
  pe012(max(), 0, 0, 0).
  
pe012(Max, Len, _I, Total) when (Len > Max) ->
  io:format("~w~n", [Total]);
    
pe012(Max, _Len, I, Total) ->
  Next = Total + I,
  Result = sqrt_factors(0, Next, 0),
  pe012(Max, Result, I + 1, Next).
    
sqrt_factors(I, N, NF) ->
  if
    ((I * I) =:= N) ->
      (2 * NF) + 1;
    ((I * I) > N) ->
      (2 * NF);
    (N rem I =:= 0) ->
      sqrt_factors(I + 1, N, NF + 1);
    true ->
      sqrt_factors(I + 1, N, NF)
  end.

File: ocaml/pe012.ml
open Core

let imax = 500 ;;

let divisible x y  = 
  x mod y = 0 ;;

let rec sqrt_factors i n nf =
  if i * i = n then
    (2 * nf) + 1
  else
    if i * i > n then
      2 * nf
    else
      if divisible n i then
        sqrt_factors (i + 1) n (nf + 1)
      else
        sqrt_factors (i + 1) n nf ;;

let rec pe012 max len i total =
  if len > max then
    printf "%d\n" total
  else
    let next = total + i in 
    let result = sqrt_factors 1 next 0 in
    pe012 max result (i + 1) next ;;

pe012 imax 0 0 0 ;;
